seqr/views/utils/terra_api_utils.py

```diff
@@ -0,0 +1,154 @@
+
+from urllib.parse import urljoin
+
+from social_django.models import UserSocialAuth
+from social_django.utils import load_strategy
+
+from settings import SEQR_VERSION, TERRA_API_ROOT_URL
+
+SEQR_USER_AGENT = "seqr/" + SEQR_VERSION
+
+logger = logging.getLogger(__name__)
+
+
+class TerraAPIException(Exception):
+    pass
+
+
+class TerraNoAccessException(TerraAPIException):
+    pass
+
+
+class TerraErrorException(TerraAPIException):
+    pass
+
+
+def anvil_enabled():
+    return bool(TERRA_API_ROOT_URL)
+
+
+def is_google_authenticated(user):
+    if not anvil_enabled():
+        return False
+    try:
+        _ = user.social_auth.get(provider = 'google-oauth2')
+    except UserSocialAuth.DoesNotExist:  # Exception happen when the user has never logged-in with Google
+        return False
+    return True
+
+
+def _get_call_args(path, headers=None, root_url=None):
+    if headers is None:
+        headers = {"User-Agent": SEQR_USER_AGENT}
+    headers.update({"accept": "application/json"})
+    if root_url is None:
+        root_url = TERRA_API_ROOT_URL
+    url = urljoin(root_url, path)
+    return url, headers
+
+
+def _get_social_access_token(user):
+    social = user.social_auth.get(provider = 'google-oauth2')
+    if (social.extra_data['auth_time'] + social.extra_data['expires'] - 10) <= int(
+            time.time()):  # token expired or expiring?
+        strategy = load_strategy()
+        try:
+            social.refresh_token(strategy)
+        except Exception as ee:
+            logger.warning('Refresh token failed. {}'.format(str(ee)))
+            raise TerraAPIException('Refresh token failed. {}'.format(str(ee)))
+    return social.extra_data['access_token']
+
+
+def anvil_call(method, path, access_token, user=None, headers=None, root_url=None, **kwargs):
+    url, headers = _get_call_args(path, headers, root_url)
+    request_func = getattr(requests, method)
+    headers.update({'Authorization': 'Bearer {}'.format(access_token)})
+    r = request_func(url, headers=headers, **kwargs)
+
+    if r.status_code == 404 or r.status_code == 403:
+        logger.warning('{} {} {} {} {}'.format(method.upper(), url, r.status_code, len(r.text), user))
+        raise TerraNoAccessException('Warning: called Terra API: /{} got status: {} with a reason: {}'.format(
+            path, r.status_code, r.reason))
+
+    if r.status_code != 200:
+        logger.error('{} {} {} {}'.format(method.upper(), url, r.status_code, len(r.text)))
+        raise TerraErrorException('Error: called Terra API: /{} got status: {} with a reason: {}'.format(
+            path, r.status_code, r.reason))
+
+    logger.info('{} {} {} {} {}'.format(method.upper(), url, r.status_code, len(r.text), user))
+
+    return json.loads(r.text)
+
+
+def _user_anvil_call(method, path, user, **kwargs):
+    access_token = _get_social_access_token(user)
+    try:
+        return anvil_call(method, path, access_token, user=user, **kwargs)
+    except TerraNoAccessException:
```

Comment:
I understand why we do this, but I'm a little worried about always ignoring all 404s and 403s. I think I would feel better if `anvil_call`  had an "allowed error" parameter that could also be passed through here, so for example in `user_get_workspace_access_level` you could do ` _user_anvil_call('get', path, user, allowed_error=403)` and then in `anvil_call` it would check if the status was an allowed error and if so return and empty dictionary and if not raise the error

Reply:
The issue is what response status codes can be silently ignored. We could have below status codes.
```
1. User has not registered (api: /register) : 404  -- not an error
2. Client ID is not whitelisted: 401 -- error
3. Access token is not valid or has expired: 401 -- error
4. Workspace does not exist or has no access permission (api: ws ACL or ws details): 404 -- not an error
5. User has access to a workspace but no ‘can share’ (api: ws ACL): 403 -- not an error
6. Other non-200 response status codes: error
```
We need to ignore both 403 and 404 for the workspace ACL requests. A single status code can't solve the issue. Maybe we should add an option to ignore both codes.

Reply:
1) If a user has not registered with anvil they should be using seqr only in local mode and google log in should have failed. This is a absolutely an error
4) If seqr is trying to access a workspace that doesn't exist something has gone really wrong and this is an error

It looks to me like the 404 cases are actually errors, so the only error code that could ever be ignored is 403. I think it also depends on what api endpoint is being called- for get workspace ACL we expect there to be 403s for some users and the behavior we want is for it to return an empty dictionary and not report any issue. For checking if a user has access to a project, if a user does not have access we expect a 403 but the check permission functions throw errors when users shouldn't have access, so  I think throwing a `PermissionDenied` error would be the easiest way to deal with it. Given that its just this one error code that has an Exception calls thats naturally associated with it, I think instead of adding `allowed_error` we should just change the error handling in `anvil_call` to throw `PermissionDenied` for a 403 and the main Terra exception for everything else. Then the `get_workspace_acl` function would have error handling just for `PermissionDenied`, and everything else wouldn't and would allow seqr to use its default error handling for that

Reply:
The `/register` API is for the detection of the user's registration on AnVIL while signing in. It is not an error because when the exception detected the code will redirect back to the login page and won't just stop running.

We can't avoid a user access a workspace that he has no permission during the transitioning phase. When a user has local access to a project has no access to the corresponding workspace (because it may take time to configure all workspaces for the seqr users), if this user requests for the workspace details or ACL then it will be a 404 status.

Maybe we can throw a TerraForbidentException for 403 and a TerraNotFoundException for 404 in `anvil_call` and let the caller function to decide how to handle them.

Reply:
"Maybe we can throw a TerraForbidentException for 403 and a TerraNotFoundException for 404 in anvil_call and let the caller function to decide how to handle them."

I like this approach. I think I would prefer to use the `PermissionDenied` exception for 403s because thats the standard exception we use all over when permission is denied so we can just let the default error handling deal with it in many cases

Reply:
Status code 403 means the access is forbidden. If we use `PermissionDenied`, I have the concern that if there is a real `PermissionDenied` during the call, it could result in miss-handling.

Reply:
permission denied means that access is forbidden, I don't understand the difference?